### DISCUSSION TITLE
BER in multi-user simulation

### DISCUSSION THREAD

**Discussion by ShelleyTimon** [NVlabs/sionna#52] [Q&A] [answered]
Hi everyone,
I just use your example Realistic_Multiuser_MIMO_Simulations to simulate the BER curve of a uplink 12-user scenario assuming UMi channel with perfect channel estimation. I set `batch_size=20` and `iter=500`. (Other parameters can be found below if you are interested.) However, the curve seems strange in two aspects. 
![batch20iter500](https://user-images.githubusercontent.com/42461121/191699284-707d5249-c851-4e52-a11d-34855f9a5c55.jpg)

1. The BER performance witnesses a sudden fluctuation when SNR is around 12-13dB.
2. After the BER decreases to 0, bit errors happen again around 18dB (since I set `early_stop=False`).
These phenomena seems quite counter-intuitive because through the simulation, BER not always decrease as SNR increases. 

Have you met such results before or will you come up with any explanation?

Thank you

Simulation parameters:
`       self._carrier_frequency = 2.6e9
        self._fft_size = 120
        self._subcarrier_spacing = 15e3
        self._num_ofdm_symbols = 14
        self._cyclic_prefix_length = 10
        self._pilot_ofdm_symbol_indices = [2, 3, 10, 11]      
        self._num_bs_ant = 64
        self._num_ut = 12
        self._num_ut_ant = 1
        self._num_bits_per_symbol = 6
        self._coderate = 2/3
    
        bs_ut_association = np.zeros([1, self._num_ut])
        bs_ut_association[0, :] = 1
        self._rx_tx_association = bs_ut_association
        self._num_tx = self._num_ut
        self._num_streams_per_tx = 1

        self._rg = ResourceGrid(num_ofdm_symbols=self._num_ofdm_symbols,
                                fft_size=self._fft_size,
                                subcarrier_spacing=self._subcarrier_spacing,
                                num_tx=self._num_tx,
                                num_streams_per_tx=self._num_streams_per_tx,
                                cyclic_prefix_length=self._cyclic_prefix_length,
                                pilot_pattern="kronecker",
                                pilot_ofdm_symbol_indices=self._pilot_ofdm_symbol_indices)

        self._sm = StreamManagement(self._rx_tx_association, self._num_streams_per_tx)
               
        self._ut_array = AntennaArray(
                                 num_rows=1,
                                 num_cols=1,
                                 polarization="single",
                                 polarization_type="V",
                                 antenna_pattern="omni",
                                 carrier_frequency=self._carrier_frequency)
        
        self._bs_array = AntennaArray(num_rows=4,
                                      num_cols=8,
                                      polarization="dual",
                                      polarization_type="cross",
                                      antenna_pattern="38.901",
                                      carrier_frequency=self._carrier_frequency)
        
        self._channel_model = UMi(carrier_frequency=self._carrier_frequency,
                                  o2i_model="low",
                                  ut_array=self._ut_array,
                                  bs_array=self._bs_array,
                                  direction="uplink",
                                  enable_pathloss=False,
                                  enable_shadow_fading=False)

        self._binary_source = BinarySource()
        self._qam_source = QAMSource(self._num_bits_per_symbol)
        
        self._n = int(self._rg.num_data_symbols*self._num_bits_per_symbol) # Number of coded bits
        self._k = int(self._n*self._coderate)                              # Number of information bits
        self._encoder = LDPC5GEncoder(self._k, self._n)
        self._decoder = LDPC5GDecoder(self._encoder)
        self._mapper = Mapper("qam", self._num_bits_per_symbol)
        self._rg_mapper = ResourceGridMapper(self._rg)
        
        self._ofdm_channel = OFDMChannel(self._channel_model, self._rg, add_awgn=True,
                                         normalize_channel=True, return_channel=True)
       
        self._remove_nulled_subcarriers = RemoveNulledSubcarriers(self._rg)
        self._ls_est = LSChannelEstimator(self._rg, interpolation_type="lin_time_avg")
        self._lmmse_equ = LMMSEEqualizer(self._rg, self._sm)
        self._demapper = Demapper("app", "qam", self._num_bits_per_symbol)`

**Comment by jhoydis** [comment, marked as answer]
The answer to your question is pretty simple: the lower the targeted BER/BLER, the larger the number of Monte Carlo iterations you need to get reliable estimates. At a BER of 1e-6, you will probably have a BLER of 1e-5 or so. If you run only 20*500=10k iterations, you get 12(=num_ut)*10k=120k simulated blocks. Thus, you will most likely observe only  a single block error from which you cannot obtain reliable BER/BLER estimates. As a rule of thumb, you should try to get at least 100 block errors for every SNR point that you are interested in. The ``sim_ber`` function shows you all of these statistics and you can easily see for which SNR points you get reliable BER/BLER estimates.

As a side remark: I think that you do simulations with perfect CSI? You would most likely not be able get into the waterfall region with 12 UTs and the Kronecker pilot pattern. This is because every UT sends a pilot on every 12th subcarrier which is probably not sufficient for the frequency selectivity of the UMI channel model. You might want to use a different pilot pattern for such scenarios.

I have run simulations with perfect CSI for the same settings as you have described with a batch size of 32 and for 1000 iterations. This is the BER result that I got (ignore the title and y-axis label of the plot which are wrong):
![image](https://user-images.githubusercontent.com/5190129/191738375-57d6b4a1-4e70-4c20-9644-d299309815d7.png)

![image](https://user-images.githubusercontent.com/5190129/191738678-5a448a8b-d41b-40b6-bd55-4115fbeec86c.png)

You can clearly see that for SNRs above 12dB the number of block errors is really small and the BER/BLER estimates are very noisy.

**Reply by ShelleyTimon** [reply]
Hi Hoydis,
thank you for your prompt answer and advice.
